### PR TITLE
Fix null repos handling for all-repos installations

### PR DESCRIPTION
**Pull request by hsubra89**
## Summary
- Preserve `null` semantics in `installations.repos` when a GitHub App is installed on all repos (previously `null` was coerced to `[]` via `?? []`, breaking `repos IS NULL` queries)
- Propagate `null` through `installation_repositories` incremental merge (all-repos installations stay null)
- Update `UpsertInstallationRequest` schema to accept `NullOr(Array(RepoEntry))`
- Add integration test for all-repos installation case
- Update spec to reflect actual implementation (incremental merge, object format, hard delete)

## Test plan
- [x] New test: `installation event with all repos → repos is null in DB`
- [x] All 25 existing integration tests still pass
- [x] `pnpm test:integration` green

Implements `specs/github-app/2-installations.md`

### CODE REVIEW

File: packages/server/src/github/webhook-handler.ts [hunk 1]
```diff
@@ -60,34 +60,32 @@ export const handleWebhook = Effect.gen(function* () {
...
-            const added: ReadonlyArray<{ full_name: string }> = payload.repositories_added ?? []
-            const existing = new Set(current.map((r) => r.full_name))
-            repos = [...current, ...added.filter((r) => !existing.has(r.full_name))]
-          } else if (action === "removed") {
-            const removed = new Set(
-              (payload.repositories_removed ?? []).map((r: { full_name: string }) => r.full_name),
-            )
-            repos = current.filter((r) => !removed.has(r.full_name))
+        } else {
+          const accountType = account?.account?.type ?? "Unknown"
```
Comment (hsyntax) [inline packages/server/src/github/webhook-handler.ts:64]: ### 🟡 WARNING (authorization)

`missing-installation-null-conflation` — `installation_repositories` handling currently treats `getInstallationRepos() === null` as "all repos," but the store uses `null` for both `repos IS NULL` and "installation row not found." A stale, replayed, or out-of-order webhook can therefore recreate a missing/deleted installation as full-account access, which is both a correctness regression and an authorization risk.

<details><summary><strong>Suggested fixes</strong></summary>

1. Change the store/API contract so missing rows are distinct from `repos = NULL` (for example, return `{ found, repos }` or a tagged union).
2. Do not upsert from `installation_repositories` when the installation row is missing; instead ignore the event, surface an error, or refetch canonical installation state before writing.
3. Add an integration test proving a deleted or absent installation is not recreated as `repos = NULL` by a later `installation_repositories` webhook.

</details>

File: packages/server/src/github/webhook-handler.ts [hunk 2]
```diff
@@ -60,34 +60,32 @@ export const handleWebhook = Effect.gen(function* () {
...
-          if (action === "added") {
-            const added: ReadonlyArray<{ full_name: string }> = payload.repositories_added ?? []
-            const existing = new Set(current.map((r) => r.full_name))
-            repos = [...current, ...added.filter((r) => !existing.has(r.full_name))]
-          } else if (action === "removed") {
-            const removed = new Set(
-              (payload.repositories_removed ?? []).map((r: { full_name: string }) => r.full_name),
-            )
-            repos = current.filter((r) => !removed.has(r.full_name))
+        } else {
```
Comment (hsyntax) [inline packages/server/src/github/webhook-handler.ts:63]: ### 🟡 WARNING (logic)

`repository-selection-all-not-handled` — The `installation_repositories` branch never uses GitHub's `repository_selection` field, so a selected-repo installation that later switches to `repository_selection: "all"` will keep an array in the database instead of being converted to SQL `NULL`. That leaves all-repo installations misrepresented and breaks `repos IS NULL` queries.

<details><summary><strong>Suggested fixes</strong></summary>

1. Short-circuit `installation_repositories` handling to write `repos = null` whenever `payload.repository_selection === "all"`.
2. Only perform incremental `repositories_added` / `repositories_removed` merges when `repository_selection === "selected"`.
3. Add an integration test that starts with a selected-repo installation, sends an `installation_repositories` webhook with `repository_selection: "all"`, and asserts the DB column becomes `NULL`.
4. Add a regression test showing an existing all-repos installation stays `NULL` after subsequent `installation_repositories` events.

</details>

**Review by hsubra89 — COMMENTED** [review]
Review: 2 finding(s) across the changes.

**Comment by hsubra89** [comment]
This PR is not merge-ready yet because repo-scope transitions can remain overbroad and the new integration tests do not match GitHub's documented `installation_repositories` payload shape.

- Verdict: ❌ `needs_fix`
- Findings: 3 total (🔴 `critical`: 0, 🟡 `warning`: 3, 🔵 `info`: 0)

### Category Breakdown
| Category | Count |
|----------|------:|
| authorization | 1 |
| quality | 1 |
| state-management | 1 |

File: packages/server/src/github/webhook-handler.ts (hunk 2 above)
Comment (hsyntax) [reply packages/server/src/github/webhook-handler.ts:63]: Fixed: repository-selection-all-not-handled: normalize all-repo installation updates

File: packages/server/src/github/webhook-handler.ts (hunk 1 above)
Comment (hsyntax) [reply packages/server/src/github/webhook-handler.ts:64]: Fixed: missing-installation-null-conflation: ignore stale installation repository webhooks

File: packages/server/src/github/webhook-handler.ts [hunk 3]
```diff
@@ -60,34 +60,38 @@ export const handleWebhook = Effect.gen(function* () {
...
+          const accountType = account?.account?.type ?? "Unknown"
+          const accountLogin = account?.account?.login ?? "unknown"
+
+          let repos: ReadonlyArray<{ full_name: string }> | null
+          if (eventType === "installation_repositories") {
+            const current = yield* store.getInstallationRepos(installationId)
```
Comment (hsyntax) [inline packages/server/src/github/webhook-handler.ts:69]: ### 🟡 WARNING (authorization)

`installation-repo-state-can-stay-overbroad` — `installation_repositories` updates currently treat both a missing installation row and an existing `repos = null` row as terminal states. That breaks legitimate state transitions: if the initial `installation` event was missed or arrives out of order, the first repo delta is dropped; and if an installation narrows from all repositories to selected repositories, `repos` can remain `NULL`, which this codebase treats as full access and can overgrant authorization after access has been revoked.

<details><summary><strong>Suggested fixes</strong></summary>

1. Track deletion separately from absence so `installation_repositories` events can rebuild state for installations that are missing but not known-deleted.
2. When `repository_selection` is `selected` and the current row is missing or has `repos = null`, replace `repos` with an authoritative selected-repo list fetched from GitHub instead of returning early or preserving `null`.
3. If an inline resync is not possible, fail closed by marking the installation as needing resync and stop treating `repos IS NULL` as authoritative until the repo list is refreshed.

</details>

File: packages/server/src/github/webhook-handler.ts [hunk 4]
```diff
@@ -60,34 +60,38 @@ export const handleWebhook = Effect.gen(function* () {
...
+              const removed = new Set(
+                (payload.repositories_removed ?? []).map((r: { full_name: string }) => r.full_name),
+              )
+              repos = current.repos.filter((r) => !removed.has(r.full_name))
+            } else {
+              repos = current.repos
+            }
           } else {
-            repos = current
+            repos = payload.repositories ?? null
```
Comment (hsyntax) [inline packages/server/src/github/webhook-handler.ts:91]: ### 🟡 WARNING (state-management)

`installation-events-can-clobber-repo-scope` — Non-`deleted` `installation` actions now write `repos = payload.repositories ?? null`, but GitHub's `installation` payload does not always include `repositories`. For actions like `suspend`, `unsuspend`, or `new_permissions_accepted`, an omitted field will rewrite a selected-repo installation to `NULL`, incorrectly broadening it to 'all repositories'.

<details><summary><strong>Suggested fixes</strong></summary>

1. Only derive `repos` from `payload.repositories` when the payload explicitly includes repository data, and preserve the stored value otherwise.
2. For non-`deleted` installation actions, load the current installation first and fall back to its existing repo set unless the payload explicitly indicates an all-repos installation.
3. Limit full repo replacement behavior to actions whose payload shape is known to carry authoritative repository scope, such as `installation.created`.

</details>

File: packages/server/tests/integration/webhook.test.ts [hunk 5]
```diff
@@ -399,6 +477,63 @@ describe("webhook receiver", () => {
     { timeout: 60_000 },
```
Comment (hsyntax) [inline packages/server/tests/integration/webhook.test.ts:477]: ### 🟡 WARNING (quality)

`installation-repositories-tests-use-invalid-payloads` — The new `installation_repositories` integration tests omit `repositories_added` and `repositories_removed`, even though GitHub documents both arrays as required for this event. That weakens coverage because the tests validate handler behavior against a synthetic payload shape rather than real deliveries.

<details><summary><strong>Suggested fixes</strong></summary>

1. Add `repositories_added: []` and `repositories_removed: []` to both new `installation_repositories` test payloads.
2. Extract a shared helper or fixture for `installation_repositories` payloads so every test includes the documented required arrays and `repository_selection` consistently.

</details>

> [!NOTE]
> This finding applies to line 314 but is shown here because that line is not in the diff.

**Review by hsubra89 — COMMENTED** [review]
Review: 3 finding(s) across the changes.
